### DISCUSSION TITLE
Help about Custom Slate Plugins

### DISCUSSION THREAD

**Discussion by GuyoST** [react-page/react-page#1151] [Q&A]
Hello every one,

I'm new to React Page and I followed the documentation to practice adding custom slate plugins.
I have 2 questions : 

### 1. about the Form Field plugin from the documentation
It tried to add the plugins shown in the doc and it was OK for the Link plugin and the Color Picker plugin, but I got trouble with the FormField plugin to add a text field.

Here is my code : 
(it's the same as the documentation, except at the end, I had to remove isVoid: true and add it just before the end, because my compiler complains about it) : 

```
import React from 'react';
import { pluginFactories } from '@react-page/plugins-slate';

const FormFieldSlatePlugin = pluginFactories.createComponentPlugin<{
    fieldName: string;
    placeholder: string;
}>({
    Component: (props) => {
        return (
            <input
                placeholder={props.placeholder}
                type="text"
                onChange={(e) =>
                    console.log(
                        'filled out field ' + props.fieldName + ', with ' + e.target.value
                    )
                }
            />
        );
    },
    controls: {
        type: 'autoform',
        schema: {
            properties: {
                fieldName: {
                    type: 'string',
                },
                placeholder: {
                    type: 'string',
                },
            },
        },
    },
    addHoverButton: true,
    addToolbarButton: true,
    type: 'FormField',
    object: 'inline',
    //isVoid: true,

    icon: <span>FormField</span>,
    label: 'FormField',
});


(FormFieldSlatePlugin as any).isVoid = true

export default FormFieldSlatePlugin
```

And then in the editor Page I added it like this : (colorPicker and linkPicker are the example from the doc, they are OK)
```
const cellPlugins = useMemo(() => [slate(def => ({
        ...def,
        plugins: {
            ...def.plugins,
            atmos: {
                colorPicker: ColorPickerSlatePlugin(),
                formField: FormFieldSlatePlugin(),
                link: LinkSlatePlugin()
            }
        }
    })), image], []);
```
When I click the form field button in the slate toolbar I get this window : 
![image](https://user-images.githubusercontent.com/4056969/152633681-7d7b3c8f-7f8f-4b49-b1ba-a2fdc537551a.png)

And when I press OK : 
![image](https://user-images.githubusercontent.com/4056969/152633688-f443ac10-9a4b-44ca-b5f9-32c73fca548d.png)

![image](https://user-images.githubusercontent.com/4056969/152633695-9e54738a-3156-4317-89e4-239424d01e2e.png)

Does someone knows why it's not working ? Did I miss something ?

### 2. general questions about slate plugins
Are there some existing plugins for the React Page Slate component (like add an image directlty, or a table...) and if they exist, where can we find them ?

Thank you very much for your help

**Comment by macrozone** [comment]
1. you have to set `isVoid; true`, otherwise you will get this error. You seem to have commented it out.

2. table does not exist yet. From my own experience in CMS projects, tables are often requested, but its often actually a bad idea to use tables for many use cases. Tables are good to show tabular data, but thats often not very nice on mobile devices. Instead of using arbitrary tables, i would ask the customer what they want to achieve to display as tables. Maybe there is a better way (e.g. with a custom cell plugin) to display that data. I would not rule out general tables in general, they are just hard to implement and hard to use and don't deliver the best UX.

images: i thought about adding images directly in slate, you can try that with a "void" plugin as shown above. Additionaly, there is a mechanism that allows cell plugins to be displayed as a floating inline element (the demo contains such an element).

**Comment by GuyoST** [comment]
Hi thank you for the information.

1. For the `isVoid`, I had to comment it out because it's not accepted otherwise, and I added it again at the end of the code with 

`(FormFieldSlatePlugin as any).isVoid = true`

even if I cast everything to any for force the compiler to accept it, I also get the error :

```
import React from 'react';
import { pluginFactories } from '@react-page/plugins-slate';

const FormFieldSlatePlugin = pluginFactories.createComponentPlugin<{
    fieldName: string;
    placeholder: string;
}>({
    Component: (props: any) => {
        return (
            <input
                placeholder={props.placeholder}
                type="text"
                onChange={(e) =>
                    console.log(
                        'filled out field ' + props.fieldName + ', with ' + e.target.value
                    )
                }
            />
        );
    },
    controls: {
        type: 'autoform',
        schema: {
            properties: {
                fieldName: {
                    type: 'string',
                },
                placeholder: {
                    type: 'string',
                },
            },
        },
    },
    addHoverButton: true,
    addToolbarButton: true,
    type: 'FormField',
    object: 'inline',
    isVoid: true,

    icon: <span>FormField</span>,
    label: 'FormField',
} as any);


export default FormFieldSlatePlugin
```

If I don't cast all as any, I can't compile : 
![image](https://user-images.githubusercontent.com/4056969/152647590-8f1b8d2b-a15c-4912-9713-4341cd22ed03.png)

So is there something to do about it ?

2. About tables of course I agree with you, it was just an example. For image that would be nice to add it in slate directly, especially to copy / paste image from the clipboard and embed them as base64 for example, or with an upload mechanism as an alternative as base64. I will have a look at the demo to see the floating image cell.

**Reply by macrozone** [reply]
1.) which version do you use? it was added very recently

2.) copy paste would be nice, but its a bit more complicated. Embedding as base64 is not a good idea generally (only for small images like svg, this is adviceable). But uploading and then saving the url is possible and the usual way to go.
